lib/Logger/LogContext.h

```diff
@@ -0,0 +1,112 @@
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Manuel Pöter
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <optional>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace arangodb {
+
+class LogContext {
+ private:
+  struct Impl;
+  
+ public:
+  /// @brief sets the key/value pair in the LogContext for the current scope.
+  /// If an entry with the specified key already exists it is overwritten,
+  /// otherwise a new key/value pair is added.
+  /// Returns a `Scoped` instance that restores the previous state. That is,
+  /// if the value of an existing entry was overwritten, the old value is
+  /// restored, otherwise the key/value pair is removed.
+  struct ScopedValue {
+    ScopedValue(std::string key, std::string value);
+    ~ScopedValue();
+  private:
+    std::size_t _idx;
+    std::optional<std::string> _oldValue;
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+    std::string _newValue;
+#endif
+  };
+
+  /// @brief merges the key/value pairs from the given context into the local
+  /// LogContext for the current scope.
+  struct ScopedMerge {
+    ScopedMerge(LogContext ctx);
+   private:
+    std::vector<ScopedValue> _values;
+  };
+
+  /// @brief sets the given LogContext as the thread's current context for the
+  /// current scope; restores the previous LogContext upon destruction.
+  struct ScopedContext {
+    ScopedContext(LogContext ctx);
+    ~ScopedContext();
+   private:
+    std::shared_ptr<Impl> _oldContext;
+  };
+  
+  LogContext();
+
+  /// @brief iterates through the local LogContext's key/value pairs and calls
+  /// the callback function for each of them.
+  void each(std::function<void(std::string const&, std::string const&)>);
+  
+  /// @brief returns the local LogContext
+  static LogContext& current();
+  
+  /// @brief sets the given context as the current LogContext.
+  static void setCurrent(LogContext ctx);
+  
+ private:
+  LogContext(std::shared_ptr<Impl> impl) : _impl(std::move(impl)) {}
+  
+  auto set(std::string key, std::string value) -> std::pair<std::size_t, std::optional<std::string>>;
+  
+  void restore(std::size_t index, std::optional<std::string>&& value);
+  
+  void ensureUniqueOwner();
+  
+  std::shared_ptr<Impl> _impl;
+};
+
+template <typename Func>
+auto withLogContext(Func func) {
+  return [func = std::move(func), ctx = LogContext::current()](auto&&... args) mutable {
+    LogContext::ScopedContext ctxGuard(std::move(ctx));
+    using result_t = std::invoke_result_t<Func, decltype(args)...>;
+    if constexpr (std::is_same_v<result_t, void>) {
+      func(std::forward<decltype(args)>(args)...);
+    } else {
+      return func(std::forward<decltype(args)>(args)...);
+    }
```

Comment:
```suggestion
auto withLogContext(Func&& func) {
  return [func = std::forward<Func>(func), ctx = LogContext::current()](auto&&... args) mutable {
    LogContext::ScopedContext ctxGuard(std::move(ctx));
    return std::forward<Func>(func)(std::forward<decltype(args)>(args)...);
```
One should call the lambda with the same ref-qualifiers as it was passed in, for example if it only callable as r-value. Therefore I added `std::forward<Func>(f)`. Furthermore you can return `void` in C++. You don't need that `if constexpr`. 

Lastly, the `LogContext` is moved out of the lambda which makes this lambda single use. This might not be true for `Func` and thus could lead to confusion and unexpected behavior.

Reply:
> One should call the lambda with the same ref-qualifiers as it was passed in, for example if it only callable as r-value. Therefore I added std::forward<Func>(f). 

Good point!

> Furthermore you can return void in C++. You don't need that if constexpr.

Unfortunately the MSVC compiler complained about that, that's why I had to add that annoying if.

> Lastly, the LogContext is moved out of the lambda which makes this lambda single use. This might not be true for Func and thus could lead to confusion and unexpected behavior.

Also a good point...

Reply:
According to [this](https://en.cppreference.com/w/cpp/language/return)
> In a function returning (possibly cv-qualified) void, the return statement with expression can be used, if the expression type is (possibly cv-qualified) void.

And [it looks like ](https://godbolt.org/z/PqoT5rWed)v19.28 doesn't complain. Which version do you use?

Reply:
If have 19.29, and for some reason it now compiled successfully. Weird..,